neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,260 @@
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
+
+## Specification
+
+The proposal is to add the following host functions to the NEAR protocol:
+
+
+```rust
+/// Instructs the protocol that the smart contract is not yet ready to respond
+/// to its caller yet.  The smart contract promises to call
+/// `promise_yield_resume()` within X (TBD) blocks.  When
+/// `promise_yield_resume()` is called, the protocol will call the method on the
+/// smart contract that is identified by `method_name_len` and `method_name_ptr`
+/// and this method may or may not respond to the caller.
+///
+/// `arguments_len` and `arguments_ptr` provide an initial blob of arguments
+/// that will be passed to the method.
+///
+/// If the contract fails to call `promise_yield_resume()` within X blocks, then
+/// the protocol will call the method with a timeout error.
+///
+/// Similar to the `gas` parameter in
+/// [promise_create](https://github.com/near/nearcore/blob/a908de36ab6f75eb130447a5788007e26d05f93e/runtime/near-vm-runner/src/logic/logic.rs#L1281),
+/// the `gas` parameter is a prepayment for the gas that would be used to
+/// execute the method.
+///
+/// `gas_weight`: as specified in
+/// [this](https://github.com/near/NEPs/blob/master/neps/nep-0264.md) NEP, this
+/// improves the devX of specifying a portion of the remaining gas for executing
+/// the method instead of specifying a precise amount.
+///
+/// `register_id`: is used to identify the register that will be used by the
+/// protocol to return unique token referring to this yielded execution to the
+/// contract.  The contract will have to pass this value when it calls
+/// `promise_yield_resume`.
+///
+/// Return value: u64: Similar to the
+/// [promise_create](https://github.com/near/nearcore/blob/a908de36ab6f75eb130447a5788007e26d05f93e/runtime/near-vm-runner/src/logic/logic.rs#L1281)
+/// host function, this function also create a promise and returns an index to
+/// the promise.  This index can be used to create a chain of promises.
+pub fn promise_yield_create(
+    method_name_len: u64,
+    method_name_ptr: u64,
+    arguments_len: u64,
+    arguments_ptr: u64,
+    gas: u64,
+    gas_weight: u64,
+    register_id: u64,
+) -> u64;
+
+/// When a smart contract has postponed replying to its caller earlier, it can
+/// use this function to indicate that it may now be ready to reply to it.  When
+/// this is called, then the protocol will call the method that the smart
+/// contract referred to in the earlier `promise_yield_create()` call.
+///
+/// `data_id_len` and `data_it_ptr`: This value was returned in an earlier call
+/// to `promise_yield_create` and uniquely identifies which yielded execution
+/// should be resumed.
```

Comment:
It should be noted that the `promise_yield_resume` may only resume delayed receipts that were created by the same contract.

(This can also be implicit by defining somewhere that these functions operate in sender-is-receiver mode only.)